### PR TITLE
Fix ComboBox text not clearing on last Item removal

### LINKED ISSUE #6621: Combobox text field is not cleared when last item is removed
.NET Core Version: 
- 7.0.0-preview.2.22102.17

Have you experienced this same bug with .NET Framework?: 
- Yes

**Problem description:**

1. Create a `Form` with a `ComboBox` and a `Button`.
2. Add two `Items` to the `ComboBox`.
3. Add an event to remove the selected `Item` when the `Button` is clicked.
4. Run the application.
5. Choose any `Item`.
6. Click on the `Button` (`Combobox` text field will be cleared).
7. Choose another `Item`.
8. Click on the `Button`.

**Actual behavior:**
The text field contains the name of the `Item`.

**Expected behavior:**
The text field should be cleared.

### PR DESCRIPTION
**Pull request by jbhensley**
Fixes #6621

## Proposed changes

`ComboBox.Text` does not clear when the last item is removed. This was tested on Win32 and confirmed to be native behavior. PR addresses by checking for last item removal and updating `Text` accordingly.

It should be noted that this likely goes back to the beginning of .NET, so this is a change to longstanding behavior. It does, however, seem to be correct when measured by the principle of least astonishment.

~~Separate issue: existing behavior does not raise `TextChanged` or `SelectedIndexChanged` when `SelectedItem` is removed. Consideration should be given as to whether this should be changed.~~

Modified to raise events. Unit tests ensure that events are raised in the correct order and only raised once. 

There is an edge which is also covered. `TextChanged` does not fire if changing between items with the same text. This includes adding an item with `string.Emtpy` as its value, selecting it, and then setting `SelectedIndex` to -1. To match that behavior, we ensure that `TextChanged` does not fire when `SelectedItem` is removed if `SelectedItem` is `string.Emtpy`.

## Customer Impact

Text will clear on last item removal, consistent with what it does when a non-last item is removed.

## Regression? 

-  No

## Risk

- It is likely that apps have been designed to compensate for this already by clearing `Text` on last item removal in an event handler. That should not create a conflict
- It is possible that apps exist which were designed to depend on this behavior, in which case it could be a breaking change. Admittedly, this would be a very odd dependency

## Test methodology 

- Unit tests added

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6673)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ObjectCollection.cs [hunk 1]
```diff
@@ -400,6 +400,11 @@ public void RemoveAt(int index)
                         _owner.UpdateText();
                     }
                 }
+                else if (InnerList.Count == 0)
```
[the remark this answers is not in the document]
Comment (jbhensley) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ObjectCollection.cs]: Agreed regarding events. Really feels like we should break with existing behavior and raise when items are removed as that would be technically correct. As is, we're just perpetuating the error. Not sure the implications for existing applications.

**Comment by SergeySmirnov-Akvelon** [comment]
@jbhensley, looks good. Great job, thank you.

@dreddy-work, should it be sent for testing to the CTI team?

**Comment by dreddy-work** [comment]
We reviewed this yesterday. It seems we should look at the associated events as well here. This is going to be a breaking change but will be documented for 7.0. @jbhensley , will you be able to look into events as well to make this new behavior complete?

**Comment by jbhensley** [comment]
> We reviewed this yesterday. It seems we should look at the associated events as well here. This is going to be a breaking change but will be documented for 7.0. @jbhensley , will you be able to look into events as well to make this new behavior complete?

Yes, I'll take a look.

**Comment by dreddy-work** [comment]
Thanks. will get this reviewed soon.

**Comment by jbhensley** [comment]
Sorry for all of the comments, but I think we should determine how to handle events when an item is inserted at index < `SelectedIndex` also. In that scenario, `SelectedIndex` will increment while `SelectedItem` and `Text` remain the same. Should `SelectedIndexChanged` be raised?

**Comment by RussKie** [comment]
Thank you for doing all these investigations, and blowing the dust off nooks and crannies. It's great to give some care to the code that has been written some 20 odd years ago, but there is a real risk of accidentally breaking consumers. who may have come to depend on existing behaviours. Please understand the team's hesitation with these kinds of changes.
We can certainly document certain existing idiosyncrasies (you can do it too from https://docs.microsoft.com/ by clicking "Edit" button in the top right hand side corner). We can also accept certain changes that may be considered breaking but have tangible benefits (e.g., a increased performance) or that aimed at addressing issues with real life scenarios.

In this case, if we can settle on the least invasive changes to achieve the goal - the better.

Hope this makes sense.

**Comment by jbhensley** [comment]
Understood. Just trying to be thorough in the analysis.

My feeling is that raising events when `SelectedItem` is removed makes sense because the selection changes. However, raising events when `SelectedIndex` changes due to an insert doesn't quite fit with what most people think the event is supposed to signify, which is a change of selection.

**Comment by Tanya-Solyanik** [comment]
@dreddy-work, @RussKie   - please remind me what our decision was, are we clearing text but are not sending any events? Or "won't fix" all aspects of this issue?

**Comment by RussKie** [comment]
I don't have a firm position - on one hand this change fixes the inconsistency, one the other hand this may result in unexpected behaviour... 
As an alternative to discarding this change - I'll be content with putting the change under a quirk (enabled by default), that can be opted-out if it breaks existing consumer.

**Comment by Tanya-Solyanik** [comment]
Quirks are hard to maintain, and this fix had not been requested by customers (at least recently), I would wait until we get a customer request.

**Comment by merriemcgaw** [comment]
I have to agree with @Tanya-Solyanik on this one - if we get a customer request we will have this fix in our back pocket. But since it will result in unexpected behavior, I am reluctant to take this without having the customer request. I also want to avoid getting in the situation where we're trying to manage dozens and dozens of quirks from proactive fixes.

**Comment by RussKie** [comment]
Hi @jbhensley, the team had a number of offline conversations discussing pros and cons, and we've concluded that the risk of an accidental regression outweighs the perceived benefits. Coupled with the fact that we don't have a real customer behind the [original issue](https://github.com/dotnet/winforms/issues/6621) (it was raised by our team member), and there is a relatively trivial workaround, we decided not to take this change. We can update the [docs](https://docs.microsoft.com/dotnet/api/system.windows.forms.combobox.text#remarks), and if you're willing you can propose an update [here](https://github.com/dotnet/dotnet-api-docs/edit/main/xml/System.Windows.Forms/ComboBox.xml).

Apologies for taking time, and thank you for your work.

**Pull request closed without merging**